[wasm][aot] Disable failing AOT tests, to get a green build

radical opened the pull request:
- this will allow tracking new failures too
- uses new `PlatformDetection.IsMonoAOT`, and `PlatformDetection.IsBuiltWithAggressiveTrimming`

radical committed:
[wasm] Disable 2 failing Common.Tests for AOT

radical committed:
[wasm][aot] Disable failing System.Collections.Tests for AOT

radical committed:
[wasm][aot] Disable failing System.Diagnostics.StackTrace.Tests for AOT

radical committed:
[wasm][aot] Disable failing System.Diagnostics.TraceSource.Tests for AOT

radical committed:
[wasm][aot] Disable 3 failing tests for  System.IO.Pipelines.Tests

radical committed:
[wasm][aot] Disable failing test for System.Memory.Tests

radical committed:
[wasm][aot] Disable failing test for System.Runtime.CompilerServices.Unsafe.Tests

radical committed:
[wasm][aot] Disable System.Text.Encodings.Web.Tests due to crash with AOT

radical committed:
[wasm][aot] Disable System.Threading.Tasks.Parallel.Tests due to timeout with AOT

radical committed:
[wasm][aot] Disable failing tests for System.Threading.Tasks.Tests

radical committed:
[wasm][aot] Disable failing tests for System.Transactions.Local.Tests

radical committed:
[wasm][aot] Disable System.Linq.Tests due to crash with AOT

radical committed:
[wasm][aot] Disable failing tests for System.CodeDom.Tests

radical commented:
cc @SamMonoRT

lewing commented:
Won't the ActiveIssue attribute disable them on the interpreter too?

SamMonoRT commented:
> Won't the ActiveIssue attribute disable them on the interpreter too?

yes, they well - @naricc is trying to clear that out via https://github.com/dotnet/runtime/pull/51522

radical committed:
[wasm] Add new PlatformDetection.IsBrowserAOT, and use that to disable

.. the failing aot tests.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-aot-green

lewing commented on eng/testing/tests.wasm.targets:
```diff
@@ -30,8 +30,10 @@
 
-    <RunScriptCommand Condition="'$(OS)' != 'Windows_NT'">$HARNESS_RUNNER $(_XHarnessArgs) %24WasmXHarnessArgs -- $(WasmXHarnessMonoArgs) %24WasmXHarnessMonoArgs $(_AppArgs) %24WasmTestAppArgs</RunScriptCommand>
-    <RunScriptCommand Condition="'$(OS)' == 'Windows_NT'">%HARNESS_RUNNER% $(_XHarnessArgs) %WasmXHarnessArgs%  -- $(WasmXHarnessMonoArgs) %WasmXHarnessMonoArgs% $(_AppArgs) %WasmTestAppArgs%</RunScriptCommand>
+    <_MonoArgs Condition="'$(RunAOTCompilation)' == 'true'">--setenv=CODE_GEN_MODE=aot</_MonoArgs>
```
Instead of doing this here, it would make more sense to put https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/mono/wasm/runtime/driver.c#L507

Using mono_eg_setenv() so that it is set in a way that doesn't depend on the runner.

radical committed:
Add `PlatformDetection.IsMonoAOT`, and update tests to use that

radical committed:
[wasm] Enable System.Runtime.CompilerServices.UnsafeTests.NullRef fixed by https://github.com/dotnet/runtime/pull/51747

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-aot-green

radical committed:
Fix missed changes

radical committed:
[wasm] Disable test failing with trimming - System.Reflection.Metadata.Tests.TagToTokenTests.ValidateTagToTokenConversion

```
[07:29:18] fail: [FAIL] System.Reflection.Metadata.Tests.TagToTokenTests.ValidateTagToTokenConversion
[07:29:18] info: System.NullReferenceException : Object reference not set to an instance of an object.
[07:29:18] info:    at System.Reflection.Metadata.Tests.TagToTokenTests.<>c__DisplayClass1_0.<GetTags>b__3()
[07:29:18] info:    at System.Reflection.Metadata.Tests.TagToTokenTests.ValidateTagToTokenConversion()
[07:29:18] info:    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

Issue: https://github.com/dotnet/runtime/issues/50714

radical committed:
[wasm] Disable individual tests failing with trimming

radical committed:
[wasm] Fix test build errors in System.Reflection.Tests with trimming

` error : FilesToIncludeInFileSystem file '/Users/radical/dev/runtime/artifacts/bin/System.Reflection.Tests/net6.0-Release/browser-wasm/publish/System.Reflection.Tests.pdb' not found`

radical committed:
[wasm] Fix System.Reflection.Tests failing due to trimming

`TestAssembly` gets trimmed causing errors like:

```
  fail: [FAIL] System.Reflection.Tests.AssemblyTests.CreateInstance_Invalid(typeName: null, exceptionType: typeof(System.ArgumentNullException))
  info: System.IO.FileNotFoundException : Could not find file '/TestAssembly.dll'.
```

lewing commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -136,6 +137,12 @@ public static bool IsNonZeroLowerBoundArraySupported
         public static bool IsOpenSslSupported => IsLinux || IsFreeBSD || Isillumos || IsSolaris;
 
+#if BUILT_WITH_EAT
+        public static bool IsBuiltWithAggressiveTrimming => true;
```
This is a trimming setting so you can set to false by default and change it to true with a linker substitution

radical committed:
Use linker substitutions to set PlatformDetection.IsBuiltWithAggressiveTrimming to true

Address review feedback from lewing

radical committed:
Use TestUtilities substitutions only if AggressiveTrimming is enabled

lewing commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -24,6 +24,7 @@ public static partial class PlatformDetection
         public static bool IsMonoRuntime => Type.GetType("Mono.RuntimeStructs") != null;
         public static bool IsNotMonoRuntime => !IsMonoRuntime;
         public static bool IsMonoInterpreter => GetIsRunningOnMonoInterpreter();
+        public static bool IsMonoAOT => !GetIsRunningOnMonoInterpreter();
```
These can both be true.

radical committed:
[wasm] Fix `PlatformDetection.IsMonoAOT`

- since we don't run with fullaot at all, both interpreter and aot
  checks should be true, when using aot-interpreter mode

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-aot-green

lewing reviewed (approved):
👍

lewing commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -404,7 +408,7 @@ private static bool GetTls13Support()
         private static bool GetIsRunningOnMonoInterpreter()
         {
-            // Browser is always using interpreter right now
+            // Browser isn't using fullaot currently
             if (IsBrowser)
                 return true;
```
```suggestion
                return RuntimeFeature.IsDynamicCodeSupported
```

lewing replied:
```suggestion
            if (IsBrowser)
                return RuntimeFeature.IsDynamicCodeSupported;
```

lewing committed:
Update src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs

lewing committed:
Update PlatformDetection.cs

radical commented:
`runtime-staging (Build windows x64 Release AllSubsets_Mono)` failing due to nuget errors, unrelated to this PR:

```
.dotnet\sdk\6.0.100-preview.2.21155.3\NuGet.RestoreEx.targets(19,5): error : (NETCORE_ENGINEERING_TELEMETRY=Restore) Failed to download package 'System.Reflection.Emit.Lightweight.4.3.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.reflection.emit.lightweight/4.3.0/system.reflection.emit.lightweight.4.3.0.nupkg'.
The SSL connection could not be established, see inner exception.
  Unable to read data from the transport connection: An existing conne
```
